spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,327 @@
+
+
+### Querying the Tuple Space
+
+The tuple space contains information regarding steps taken by validators during possibly many rounds of possibly many heights.
+Each entry has form $\lang Height, Round, Step, Validator, Value \rang$ and corresponds to the message Validator sent in Step of Round of Height; Value is a tuple of the message contents.
+
+A query to the tuple space has the same form as the entries, with parts replaced by values, that must match the values in the entries, or by `*`, which matches any value.
+For example, suppose the tuple space has the following entries, here organized as rows of a table for easier visualization:
+
+| Height | Round | Step     | Validator | Value |
+|--------|-------|----------|-----------|-------|
+| H      | R     | Proposal | v         | pval  |
+| H      | R     | PreVote  | v         | vval  |
+| H      | R'    | PreCommit| v'        | cval  |
+| H'     | R     | Proposal | v         | pval' |
+| H'     | R''   | PreVote  | v'        | vval' |
+| H'     | R'''  | PreCommit| v'        | cval' |
+
+- Query $\lang H, R, Proposal, v, * \rang$ returns $\{ \lang H, R, Proposal, v, pval \rang \}$
+- Query $\lang H, R, *, v, * \rang$ returns $\{ \lang H, R, Proposal, v, pval \rang,  \lang H, R, PreVote, v, vval \rang \}$.
+
+
+### When to remove
+
+> **TODO**: Expand on when to remove, based on the algorithm.
+
+
+### Local views
+
+The tuple space information is distributed among nodes that keep local views of the whole space.
+Because of the asynchronous nature of the system, $t_p$ may not include entries in the space or may still include entries no longer in the space.
+Formally, let $T$ be the tuple space and $t_p$ be node $p$'s view of $T$; $T = \cup_p t_p$.
+
+- $e \in T \Leftrightarrow \exists p, e \in t_p \land \not\exists q, \bar{e}\in t_q$
+
+Nodes can only query local views, not $T$.
+Queries are subscripted with the node being queried.
+
+
+### State Validity
+
+Given that each validator can execute each step only once per round, a query that specifies height, round, step and validator must either return empty or a single tuple.
+
+- $\forall h \in \N, r \in \N, s \in \text{Proposal, PreVote, PreCommit}, v \in \text{ValSet}_{h,r}$,  $\lang h, r, s, v, * \rang$ returns at most one value.
+
+In the specific case of the Proposal step, only the proposer of the round can have a matching entry.
+
+- $\lang H, R, Proposal, *, * \rang$ returns at most one value.
+
+A violation of these rules is a proof of misbehavior.
+
+
+### Convergence Requirements
+
+We now formalize the requirements of the Tendermint algorithm in terms of an eventually consistent Tuple Space, which may be implemented using reliable communication (Gossip Communication), some best *best-effort* communication primitive that only delivers messages that are still useful, or some Gossip/Epidemic/Anti-Entropy approach for state convergence.
```

Comment:
It can be confusing here to mention 3 communication mechanisms for ensuring convergence (reliable/gossip, best-effort, epidemic). While reading this, I feel like there is an insight here behind the reason for mentioning all of them, but I don't get the insight. Can you clarify, or simplify?